File: frontend/src/api/attackGraph.ts
import api from './client'

export interface AttackGraphData {
  case_id:    string
  nodes:      Record<string, unknown>[]
  edges:      Record<string, unknown>[]
  updated_at: string
}

export const attackGraphApi = {
  get: (caseId: string): Promise<AttackGraphData> =>
    api.get<AttackGraphData>(`/cases/${caseId}/attack-graph`).then(r => r.data),

  save: (caseId: string, nodes: unknown[], edges: unknown[]): Promise<AttackGraphData> =>
    api.put<AttackGraphData>(`/cases/${caseId}/attack-graph`, { nodes, edges }).then(r => r.data),

  /**
   * The graph as a PNG, rendered server-side by the same code the DOCX report
   * uses — so the download and the report image are the same picture.
   */
  png: (caseId: string): Promise<Blob> =>
    api.get(`/cases/${caseId}/attack-graph/png`, { responseType: 'blob' }).then(r => r.data),

  /**
   * Store a PNG of the canvas so the report embeds what the analyst arranged
   * rather than a server-side redrawing of it.
   */
  saveSnapshot: (caseId: string, png: Blob): Promise<void> =>
    api.put(`/cases/${caseId}/attack-graph/snapshot`, png, {
      headers: { 'Content-Type': 'image/png' },
    }).then(() => undefined),
}

File: backend/app/models/attack_graph.py
from datetime import UTC, datetime

from sqlalchemy import JSON, Column, DateTime, ForeignKey, LargeBinary, String

from app.database import Base


class AttackGraph(Base):
    """One attack graph per case — case_id is the PK."""
    __tablename__ = "attack_graphs"

    case_id    = Column(String, ForeignKey("cases.id", ondelete="CASCADE"), primary_key=True)
    nodes      = Column(JSON, nullable=False, default=list)
    edges      = Column(JSON, nullable=False, default=list)
    updated_at = Column(
        DateTime,
        default=lambda: datetime.now(UTC),
        onupdate=lambda: datetime.now(UTC),
        nullable=False,
    )

    # A PNG of the canvas, rasterised by the browser that drew it.
    #
    # The server can render this graph from the coordinates above, and does when
    # no snapshot exists — but that is a redrawing, and a redrawing never quite
    # matches the screen. Storing what the analyst actually saw means the report
    # embeds the picture they arranged, not an approximation of it.
    snapshot_png = Column(LargeBinary, nullable=True)
    snapshot_at  = Column(DateTime, nullable=True)
